crates/accelerate/src/unitary_synthesis.rs

```diff
@@ -111,16 +111,31 @@ fn apply_synth_dag(
     out_qargs: &[Qubit],
     synth_dag: &DAGCircuit,
 ) -> PyResult<()> {
+    let mut out_dag_concat = out_dag.as_concat();
     for out_node in synth_dag.topological_op_nodes()? {
```

Comment:
This function (`apply_synth_dag`) doesn't feel like it's high enough in the call tree, because we're allocating and de-allocating the concatenator for each synthesised unitary matrix, whereas in an ideal world we'd do that only once. Perhaps we should be passing the concatenator through the function calls?

Reply:
The call to `out_dag.as_concat()` was moved higher up in the stack in bf26f56.